pilz_control/include/pilz_control/traj_mode_manager.h

```diff
@@ -0,0 +1,249 @@
+/*
+ * Copyright (c) 2020 Pilz GmbH & Co. KG
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#ifndef TRAJPROCESSINGMODEMANAGER_H
+#define TRAJPROCESSINGMODEMANAGER_H
+
+#include <array>
+#include <list>
+#include <memory>
+#include <mutex>
+#include <condition_variable>
+
+namespace pilz_joint_trajectory_controller
+{
+
+enum class TrajProcessingMode
+{
+  //! The controller accepts new trajectories.
+  unhold,
+  //! The controller rejects all new incoming trajectories and stops the robot.
+  stopping,
+  //! The robot stands still and the controller rejects all new incoming trajectories.
+  hold
+};
+
+// "Simple" linear state machine: (start with stopping in accordance to the controller)
+//  x ->  |->  stopping  ->  hold  ->  unhold ->|
+//        |<-               <-                <-|
+static constexpr unsigned int NUM_MODES {3};
+static constexpr std::array<TrajProcessingMode, NUM_MODES> mode_state_machine_{
+  TrajProcessingMode::stopping, TrajProcessingMode::hold, TrajProcessingMode::unhold
+};
+
+inline static constexpr TrajProcessingMode getMode(const unsigned int& idx)
+{
+  return mode_state_machine_[idx];
+}
+
+inline static constexpr bool isTransitionValid(const TrajProcessingMode& requested_new_mode,
+                                               const unsigned int& idx_of_request_new_mode)
+{
+  return requested_new_mode == getMode(idx_of_request_new_mode);
+}
+
+inline static constexpr unsigned int getNextIndex(const unsigned int& current_idx)
+{
+  return (current_idx+1) % NUM_MODES;
+}
+
+/**
+ * @brief Listener to wait for a specified mode to be reached.
+ */
+class TrajProcessingModeListener
+{
+public:
+   //! @param mode Mode to wait for.
+  TrajProcessingModeListener(const TrajProcessingMode& mode);
+  void waitForMode();
+  //! @returns true if the given mode corresponds to the target mode, otherwise false.
+  bool isTargetModeReached(const TrajProcessingMode& mode) const;
+  void triggerListener();
```

Comment:
Done.

Reply:
I would suggest to rename `isTargetModeReached`->`isTargetMode`, because that describes exactly method does. The name `isTargetModeReached` originates from the context where the method is called, which makes it hard to understand. What do you think?

Reply:
Done.